voxels/Voxelize.cpp

```diff
@@ -371,13 +279,514 @@ std::vector<std::byte> raw_voxelize_obj(std::string_view filepath,
+                            ss << (directory + diffuse_texname);
+                            ss << "\n";
+                            ASSERT(false, ss.str());
+                        }
+                        loaded_textures_[face_mat_id] = {
+                            texture_pixels, texture_width, texture_height};
+                    }
+                }
+            }
+        }
+    }
+}
+
+Model::~Model() {
+    for (auto [_, tex] : loaded_textures_) {
+        stbi_image_free(std::get<0>(tex));
+    }
+}
+
+Voxelizer::Voxelizer(std::string_view filepath, float voxel_size) : 
+    model_{filepath}, current_memory_usage_{0} 
+{
+    glm::vec3 max = model_.get_max();
+    glm::vec3 min = model_.get_min();
+    width_ = static_cast<uint32_t>(ceil((max.x - min.x) / voxel_size));
+    height_ = static_cast<uint32_t>(ceil((max.y - min.y) / voxel_size));
+    depth_ = static_cast<uint32_t>(ceil((max.z - min.z) / voxel_size));
+    std::cout << "Chunk size: (" 
+              << width_ << ", " << height_ << ", " << depth_ 
+              << ")\n";
+
+    // Allocate the number of voxel chunks that we'll need.
+    chunks_width_ = static_cast<uint32_t>(ceil((float)width_ / voxel_chunk_size_));
+    chunks_height_ = static_cast<uint32_t>(ceil((float)height_ / voxel_chunk_size_));
+    chunks_depth_ = static_cast<uint32_t>(ceil((float)depth_ / voxel_chunk_size_));
+
+    voxel_chunks_.resize(chunks_width_ * chunks_height_ * chunks_depth_);
+
+    std::stringstream string_pointer;
+    string_pointer << this;
+    voxels_directory_ = "disk_voxels_" + string_pointer.str();
+    std::filesystem::create_directory(voxels_directory_);
+}
+
+Voxelizer::Voxelizer(std::vector<std::byte> &&voxels, uint32_t width, uint32_t height, uint32_t depth) : 
+    width_{width},
+    height_{height},
+    depth_{depth}
+{
+    voxel_chunks_.push_back(voxels);
+}
+
+Voxelizer::~Voxelizer() {
+    std::filesystem::remove_all(voxels_directory_);
+}
+
+inline uint32_t Voxelizer::linearize_chunk_index(uint32_t x, uint32_t y, uint32_t z) const {
+    return x + chunks_width_ * y + chunks_width_ * chunks_height_ * z;
+}
+
+inline std::tuple<uint32_t, uint32_t, uint32_t> Voxelizer::get_voxel_chunk_index(
+    uint32_t x, uint32_t y, uint32_t z) const {
+    uint32_t chunk_x = static_cast<uint32_t>(floor((float)x / voxel_chunk_size_)); 
+    uint32_t chunk_y = static_cast<uint32_t>(floor((float)y / voxel_chunk_size_));
+    uint32_t chunk_z = static_cast<uint32_t>(floor((float)z / voxel_chunk_size_));
+
+    return {chunk_x, chunk_y, chunk_z};
+}
+
+void Voxelizer::voxelize_chunk(uint32_t chunk_x, uint32_t chunk_y, uint32_t chunk_z) {
+    // If this function is being called, we can assume that the chunk isn't already in memory.
+    // We first check if the chunk instead exists on the disk.
+    uint32_t voxel_chunk_index =
+        linearize_chunk_index(chunk_x, chunk_y, chunk_z);
+    VoxelizedChunk &voxel_chunk = voxel_chunks_[voxel_chunk_index];
```

Comment:
Given you're encoding the voxel chunk map as a vector, I'd recommend using .at() to make sure there's no funny business